lib/src/native/realm_core.dart

```diff
@@ -500,52 +502,45 @@ class _RealmCore {
 
     if (syncError is ClientResetError) {
       syncConfig.clientResetHandler.onManualReset?.call(syncError);
-    } else {
-      syncConfig.syncErrorHandler(syncError);
+      return;
+    }
+
+    syncConfig.syncErrorHandler(syncError);
+  }
+
+  static void _callResetCallback(FutureOr<void> Function() callback, Pointer<Void> unlockCallbackFunc) async {
+    bool success = true;
+    try {
+      await callback();
+    } catch (error) {
+      success = false;
+      _realmLib.realm_register_user_code_callback_error(error.toPersistentHandle());
+    } finally {
+      _realmLib.realm_dart_invoke_unlock_callback(success, unlockCallbackFunc);
     }
   }
 
   static void _syncBeforeResetCallback(Object userdata, Pointer<shared_realm> realmHandle, Pointer<Void> unlockCallbackFunc) {
     final syncConfig = userdata as FlexibleSyncConfiguration;
-    final beforeResetCallback = syncConfig.clientResetHandler.onBeforeReset;
-    if (beforeResetCallback == null) {
-      _invokeNativeFunction(unlockCallbackFunc);
-      return;
-    }
+    final beforeResetCallback = syncConfig.clientResetHandler.onBeforeReset!;
```

Comment:
I wouldn't rely on the implementation of another function to make sure that onBeforeReset is not null. We can change it after two version and to forget about that requirement. 
What bothers me is that if some error happens inside `_syncBeforeResetCallback` method we will never return and will hold C thread locked. It is good to have `try` block and in its `catch` to always unlock the C thread with `call_invokeNativeFunction(unlockCallbackFunc);`. An error could happen also when calling `RealmInternal.getUnowned`. It should be also in the `try`.
The same is for `_syncAfterResetCallback`

Reply:
That is exactly the idea and what it should do. This function  *demands* that this is not null if it is invoked. If it is it will throw an exception. That is the idea of the hard cast.

Reply:
As Desi points out, if it is null it will lead to a deadlock because we're doing this outside of the try block. Additionally, it's kind of dangerous if due to a bug in our SDK, users start losing data.

Reply:
If you try to `throw Exception("test exception");` here, we will have a deadlock.

Reply:
then the previous code was not written correctly either since we are casting the config. Which might fail as well.

Reply:
I will be fixing this

Reply:
fixed